Adds classban module providing extban 'n' for connect class ban.

johanna-a opened the pull request:
This adds a module, connameban, that provides extban token 'n' that matches the mask against the users ->GetClass()->name.
This is not yet tested in production.

The IS_LOCAL() should work in all cases that are contained in master, since in those cases only local users are handled. Changing the signature of OnCheckBan to take a LocalUser* instead of a User* could make sense.

If not, I guess adding some sort of handling of the case where the user is not local would be necessary in this module, I'm not sure how to get that information across server nodes though.

attilamolnar commented on src/modules/m_connameban.cpp:
```diff
@@ -0,0 +1,56 @@
+
+class ModuleConnameBan : public Module, public Whois::EventListener
+{
+	public:
```
Please only indent this with one space.

attilamolnar commented on src/modules/m_connameban.cpp:
```diff
@@ -0,0 +1,56 @@
+class ModuleConnameBan : public Module, public Whois::EventListener
+{
+	public:
+		Version GetVersion() CXX11_OVERRIDE
```
Indent with one tab

attilamolnar commented on src/modules/m_connameban.cpp:
```diff
@@ -0,0 +1,56 @@
+			return MOD_RES_PASSTHRU;
+		}
+
+		void OnWhois(Whois::Context& whois) CXX11_OVERRIDE
```
I don't think coupling the extban with a whois line is a great idea, let's keep this simple and only have the extban and we'll figure out how's best to show the connect class in whois.

attilamolnar commented on src/modules/m_connameban.cpp:
```diff
@@ -0,0 +1,56 @@
+
+		ModuleConnameBan() : Whois::EventListener(this) {};
+
+		ModResult OnCheckBan(User *user, Channel *c, const std::string& mask) CXX11_OVERRIDE
```
The `*` (and `&`) should be next to the type, not the variable name.

attilamolnar commented on src/modules/m_connameban.cpp:
```diff
@@ -0,0 +1,56 @@
+	public:
+		Version GetVersion() CXX11_OVERRIDE
+		{
+			return Version("Conname 'n' - Connection name (class) ban",VF_OPTCOMMON);
```
The flags should be `VF_VENDOR | VF_OPTCOMMON` as it's going to be part of the official distribution.

SaberUK commented:
This module needs to also be documented in docs/conf/modules.conf.example.

Bikeshed: I think this module would be better off named `m_classban` as that shows its intention better.

attilamolnar commented:
Thanks for this module, I've commented on the diff (mostly about the coding style).

I agree that the module should be renamed as most people know connect classes as connect classes.

johanna-a commented:
Thank you for your feedback :)
I've updated the PR and squashed the commits.
Since we're separating class info in whois from this module perhaps it should be included in core_whois? I guess it might be relevant information to opers, but probably not to regular users.

SaberUK commented on src/modules/m_classban.cpp:
```diff
@@ -0,0 +1,49 @@
+	ModResult OnCheckBan(User* user, Channel* c, const std::string& mask) CXX11_OVERRIDE
+	{
+		LocalUser* localUser = IS_LOCAL(user);
+		if (localUser && (mask.length() > 2) && (mask[0] == 'n') && (mask[1] == ':'))
```
Bikeshed: 'c' might be a better extban character than 'n' here.

attilamolnar replied:
Put the first check (localuser) in parenthesis as well to avoid inconsistency with the other checks in this if

attilamolnar commented on src/modules/m_classban.cpp:
```diff
@@ -0,0 +1,49 @@
+		return Version("Class 'n' - Connection class ban", VF_VENDOR | VF_OPTCOMMON);
+	}
+
+	ModuleClassBan() {};
```
Empty, can remove

attilamolnar commented on src/modules/m_classban.cpp:
```diff
@@ -0,0 +1,47 @@
+class ModuleClassBan : public Module
+{
+ public:
+	Version GetVersion() CXX11_OVERRIDE
```
Please move this method to the bottom of the class to be consistent with the rest of the modules

attilamolnar commented on docs/conf/modules.conf.example:
```diff
@@ -464,6 +464,11 @@
+#
+#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#
+# Connection class ban module: Adds support for extban 'n' which 
+# matches against the class name of the user's connection.
```
I think we should mention that for this module to work properly on a network with multiple servers, all servers on the network should refer to the same class with the same name. For example, if there is a webirc class on one server and not on another, it won't work.

attilamolnar replied:
What do you think? Feel free to suggest better wording.

attilamolnar commented:
Thanks!
